compiler/rustc_infer/src/infer/generalize.rs

```diff
@@ -139,6 +153,12 @@ struct Generalizer<'me, 'tcx, D> {
 
     cache: SsoHashMap<Ty<'tcx>, Ty<'tcx>>,
 
+    /// This is set once we're generalizing the arguments of an alias. In case
+    /// we encounter an occurs check failure we generalize the alias to an
+    /// inference variable instead of erroring. This is necessary to avoid
+    /// incorrect errors when relating `?0` with `<?0 as Trait>::Assoc`.
+    in_alias: bool,
```

Comment:
exists because of `<() as Trait<<?0 as Trait<()>>::Assoc>>::Assoc == ?0` have to update comment. With `Projection` goals the behavior is still broken, but I want to change it to emit `alias-relate` in the new solver in a different PR, at which point this should work.